<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Shiny App • chevreulShiny</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Shiny App">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">chevreulShiny</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.26</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/chevreulShiny.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/shiny_app.html">Shiny App</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cobriniklab/chevreulShiny/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Shiny App</h1>
                        <h4 data-toc-skip class="author">Kevin
Stachelek</h4>
            <address class="author_afil">
      University of Southern
California<br><a class="author_email" href="mailto:#"></a><a href="mailto:kevin.stachelek@gmail.com" class="email">kevin.stachelek@gmail.com</a>
      </address>
                              <h4 data-toc-skip class="author">Bhavana
Bhat</h4>
            <address class="author_afil">
      University of Southern
California<br><a class="author_email" href="mailto:#"></a><a href="mailto:bbhat@usc.edu" class="email">bbhat@usc.edu</a>
      </address>
                  
            <h4 data-toc-skip class="date">22 March 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/cobriniklab/chevreulShiny/blob/master/vignettes/shiny_app.Rmd" class="external-link"><code>vignettes/shiny_app.Rmd</code></a></small>
      <div class="d-none name"><code>shiny_app.Rmd</code></div>
    </div>

    
    
<p>chevreulShiny allows exploratory data analysis of a single cell
dataset processed chevreulShiny
ngleCellExperiment<code>with the command</code>minimalSceApp()`</p>
<p>Here we’ll provide a brief tour of this app organized by section with
app functions organized as tabs along the sidebar upon startup.</p>
<div class="section level3">
<h3 id="reformat-metadata">Reformat Metadata<a class="anchor" aria-label="anchor" href="#reformat-metadata"></a>
</h3>
<p>Arbitrary variables can be appended to the cell metadata based on the
results of exploratory data analysis.</p>
<p>Metadata addition can be executed by uploading a csv with row names
matching cell ids with new variables as columns. Metadata editing can
also be accomplished using a built-in spreadsheet tool.</p>
</div>
<div class="section level3">
<h3 id="plot-data">Plot Data<a class="anchor" aria-label="anchor" href="#plot-data"></a>
</h3>
<div class="section level4">
<h4 id="dimensional-reduction-plots">Dimensional reduction plots<a class="anchor" aria-label="anchor" href="#dimensional-reduction-plots"></a>
</h4>
<p>chevreulShiny provides visualization of embeddings in PCA, tSNE, and
UMAP for scRNAseq data summarized at the gene and transcript level.
Plots can be customized to display cell metadata and the results of
community detection clustering by Louvain or Leiden algorithms as
implemented in SingleCellExperiment. In addition, gene and transcript
feature expression can be overlaid on embedding plots both at the gene
and transcript level.</p>
<p><img src="images%2Flouvain_and_feature_umap_plot.png"></p>
</div>
<div class="section level4">
<h4 id="readumi-count-histograms">read/UMI count histograms<a class="anchor" aria-label="anchor" href="#readumi-count-histograms"></a>
</h4>
<p>To facilitate quality control, cell-level summarized UMI and read
count values can also be overlaid with cell metadata and clustering
results.</p>
<p><img src="images%2Fpair_of_histograms.png"></p>
</div>
<div class="section level4">
<h4 id="clustering-trees">Clustering trees<a class="anchor" aria-label="anchor" href="#clustering-trees"></a>
</h4>
<p>A cluster tree of cell identities is displayed using <a href="https://lazappi.github.io/clustree/articles/clustree.html" class="external-link">scclustviz</a>.
This plot can be used to help establish an optimal clustering
resolution.</p>
<p>In order to use this function <code><a href="https://github.com/lazappi/clustree" class="external-link">library(clustree)</a></code> must be
loaded.</p>
<p><img src="images%2Fclustree.png"></p>
</div>
</div>
<div class="section level3">
<h3 id="heat-mapsviolin-plots">Heat Maps/Violin Plots<a class="anchor" aria-label="anchor" href="#heat-mapsviolin-plots"></a>
</h3>
<div class="section level4">
<h4 id="heat-maps">
<em>Heat Maps</em><a class="anchor" aria-label="anchor" href="#heat-maps"></a>
</h4>
<p>When plotting the heat map, gene/transcript expression values for
each cell are normalized by that cell’s total expression then multiplied
by 10,000 and natural-log transformed before plotting.</p>
<p>Clustering model or metadata for column arrangement is taken as a
parameter to produce a complex heat map.</p>
<p>By default, 50 most highly variable genes are displayed. However, an
arbitrary lists of genes can be plotted for comparison. The
genes/transcripts are displayed in the order they are listed.</p>
<p><img src="images%2FHeatmap2.0.png"></p>
</div>
<div class="section level4">
<h4 id="violin-plots">
<em>Violin Plots</em><a class="anchor" aria-label="anchor" href="#violin-plots"></a>
</h4>
<p>Feature (gene or transcript) expression can be viewed by violin plot
based on sample metadata</p>
<p>Violin plots are a hybrid of density and box plots. It shows the
probability density of the data at different values providing a visual
representation of distribution of expression level of a feature in
different groups of cell. The horizontal line is a marker for the median
of the data and the box shows the interquartile ranges.</p>
<p>The parameters that could to be chosen here are: 1. ‘Grouping
variable’ 2. ‘Data Type (Transformed/Raw)’</p>
<p><img src="images%2FviolinPlot.png"></p>
</div>
</div>
<div class="section level3">
<h3 id="differential-expression">Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h3>
<p>chevreulShiny provides different methods of differential expression
analysis that users can choose from to determine differential expression
of gene. DE methods included in chevreulShiny are t-test, Wilcoxon
rank-sum test, and pairwise binomial test.</p>
<p>Running DE testing results in a data frame containing the following
information: * p_val : Unadjusted p-value * avg_log2FC : log fold-change
of the average expression between the two groups. * pct.1 : The
percentage of cells where the feature is detected in the first group *
pct.2 : The percentage of cells where the feature is detected in the
second group * p_val_adj : Adjusted p-value, based on bonferroni
correction using all features in the dataset.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"Dominic.csv"</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="volcano-plots">
<em>Volcano Plots</em><a class="anchor" aria-label="anchor" href="#volcano-plots"></a>
</h4>
<p>The results of differential expression analyses can be visualized
using a volcano plot. This plot helps in visual identification of genes
with large fold changes that are also statistically significant.</p>
<p><img src="images%2FvolcanoPlot.png"> ### Find Markers</p>
<p>Marker features based on louvain/leiden cluster identities or cell
metadata can be defined based on results of wilcoxon rank-sum test.
Variable numbers of marker genes per cell group can be specified based
on adjusted p value and thresholded log fold change.</p>
<p><img src="images%2Fmarker_gene_plot.png"></p>
</div>
</div>
<div class="section level3">
<h3 id="subset-singlecellexperiment-input">Subset SingleCellExperiment Input<a class="anchor" aria-label="anchor" href="#subset-singlecellexperiment-input"></a>
</h3>
<p>It is often useful to subset a single cell data set based on cell
metadata whether experimentally determined (age, collection method,
etc.) or derived from analysis (quality control metrics, annotated cell
type)</p>
<p>chevreulShiny makes it simple to subset a dataset consisting of a
single batch or batch-integrated dataset. Subsetting can be accomplished
either in a graphical setting by lasso-selection from a dimensionally
reduced plot or by by specification of a formatted .csv file.</p>
<p>Subsetting of single or batch integrated data will trigger renewal of
all relevant preprocessing steps including dimensional reduction,
clustering, marker gene, and pathway enrichment as well as integration
based on a ‘batch’ variable</p>
</div>
<div class="section level3">
<h3 id="all-transcripts">All Transcripts<a class="anchor" aria-label="anchor" href="#all-transcripts"></a>
</h3>
<p>If expression is summarized with both gene and transcript
experiments, it is possible to plot all constituent transcripts (labeled
by ensembl transcript ids) making up a given gene.</p>
<p>The two parameters that must be chosen are:</p>
<ol style="list-style-type: decimal">
<li>dimensional reduction method: PCA, UMAP or tSNE</li>
<li>The name of the gene of interest</li>
</ol>
<p><img src="images%2Funnamed-chunk-9-1.png"></p>
<p>Direct comparison of the contribution of individual transcripts can
be achieved using stacked bar plots to answer the question: What
contribution does each transcript make for each gene expression?</p>
<p><img src="images%2Ftranscript_composisiont.png"></p>
</div>
<div class="section level3">
<h3 id="regress-features">Regress Features<a class="anchor" aria-label="anchor" href="#regress-features"></a>
</h3>
<p>To correct for expression variation in analysis due to phenomena
unrelated to focus of the study, rather than excluding count values
attributable to such process, we can adjust the expression of all
remaining genes or transcripts in each cell based on the sum score of
relevant genes, to regress out cell-cycle effects.</p>
</div>
<div class="section level3">
<h3 id="technical-information">Technical information<a class="anchor" aria-label="anchor" href="#technical-information"></a>
</h3>
<p>Technical information regarding upstream processing, software version
numbers, and dataset features can be viewed.</p>
</div>
<div class="section level2">
<h2 id="server-mode">Server Mode<a class="anchor" aria-label="anchor" href="#server-mode"></a>
</h2>
<p>If we want work with objects interactively and save the results of
our exploratory data analysis we need to operate chevreulShiny in
<em>server mode</em>. When operating in this manner it is possible to
carry out basic project management tasks including: * Loading saved
objects * Saving objects after making changes in the app such as : *
Reformatting metadata * Subsetting to remove cells * Regressing feature
expression as well as * Integrating multiple saved objects</p>
<p>Finally, server mode is necessary for some expanded functionality
including: * Coverage plotting</p>
<p><em>server mode</em> requires access to a few sqlite databases with
file path details stored in <code>~/.cache/chevreul</code>.
<code>single_cell_projects.db</code> and <code>bigwig_files.db</code>
allowing dynamic loading of objects and bigwig (.bw) sample coverage
files for each cell. These are initialized upon package installation
using the <code>creat_project_db</code> function.</p>
<p>The following sections will give details on this expanded
functionality.</p>
<div class="section level3">
<h3 id="integrate-projects">Integrate Projects<a class="anchor" aria-label="anchor" href="#integrate-projects"></a>
</h3>
<p>While operating chevreulShiny in server mode, separate sequencing
batches can be <a href="https://www.cell.com/cell/fulltext/S0092-8674(19)30559-8" class="external-link">integrated</a>
and compared to validate sample processing steps and exclude technical
variation in favor of relevant biological variation.</p>
<p><img src="images%2Fintegration_result.png" style="width:80.0%;height:80.0%"></p>
<p>This section includes a list of projects that can be selected,
integrated, and saved for future analysis.</p>
</div>
<div class="section level3">
<h3 id="coverage-plots">Coverage plots<a class="anchor" aria-label="anchor" href="#coverage-plots"></a>
</h3>
<p>Fine-grained analysis of isoform contributions can be achieved by
plotting absolute read coverage plots across a given gene.</p>
<p>indicate the **read depths across all transcripts for different
groups of cells within a genomic region.</p>
<p>The three user input parameters are: 1. ‘Select a gene’- Select a
gene of interest 2. ‘Color by variable’- Select a variable by which to
group the cells by 3. ‘Groups to display’- Select the groups to be
displayed</p>
<p><img src="images%2FPlotCoverage2.0.png" style="width:80.0%;height:80.0%"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.4.3 (2025-02-28)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8        LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [6] LC_MESSAGES=C.UTF-8    LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C           LC_TELEPHONE=C        </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] sessioninfo_1.2.3 BiocStyle_2.34.0 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] digest_0.6.37       desc_1.4.3          R6_2.6.1            bookdown_0.42       fastmap_1.2.0      </span></span>
<span><span class="co">#&gt;  [6] xfun_0.51           cachem_1.1.0        knitr_1.50          htmltools_0.5.8.1   rmarkdown_2.29     </span></span>
<span><span class="co">#&gt; [11] lifecycle_1.0.4     cli_3.6.4           sass_0.4.9          pkgdown_2.1.1       textshaping_1.0.0  </span></span>
<span><span class="co">#&gt; [16] jquerylib_0.1.4     systemfonts_1.2.1   compiler_4.4.3      tools_4.4.3         ragg_1.3.3         </span></span>
<span><span class="co">#&gt; [21] evaluate_1.0.3      bslib_0.9.0         yaml_2.3.10         BiocManager_1.30.25 jsonlite_1.9.1     </span></span>
<span><span class="co">#&gt; [26] rlang_1.1.5         fs_1.6.5            htmlwidgets_1.6.4</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kevin Stachelek, Bhavana Bhat.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
